pkg/operator/controller/dns/controller.go

```diff
@@ -222,13 +223,26 @@ func (r *reconciler) createDNSProviderIfNeeded(dnsConfig *configv1.DNS) error {
 	creds := &corev1.Secret{}
 	switch platformStatus.Type {
 	case configv1.AWSPlatformType, configv1.AzurePlatformType, configv1.GCPPlatformType:
-		name := types.NamespacedName{Namespace: r.config.Namespace, Name: cloudCredentialsSecretName}
-		if err := r.cache.Get(context.TODO(), name, creds); err != nil {
-			return fmt.Errorf("failed to get cloud credentials from secret %s: %v", name, err)
+		if (r.infraConfig != nil && infraConfig.Status.ControlPlaneTopology != configv1.ExternalTopologyMode) || r.infraConfig == nil {
```

Comment:
The condition can be simplified:
```suggestion
		if r.infraConfig == nil || infraConfig.Status.ControlPlaneTopology != configv1.ExternalTopologyMode {
```
But really, why check `r.infraConfig` at all in this `switch` statement?  If the condition `r.infraConfig == nil` is true, then that means that the `if r.infraConfig == nil || ...` statement below will set `needUpdate = true`, and if `infraConfig.Status.ControlPlaneTopology == configv1.ExternalTopologyMode` is true, then when we call `createDNSProvider`, it will return without using `creds`.  So I think what you had in the previous version of this PR is sufficient: 
```suggestion
		if infraConfig.Status.ControlPlaneTopology != configv1.ExternalTopologyMode {
```